File: src/main/java/ltdjms/discord/shared/cache/CacheKeyGenerator.java
package ltdjms.discord.shared.cache;

/**
 * 緩存鍵生成器介面。
 *
 * <p>提供統一的緩存鍵格式，避免鍵衝突並便於管理。 鍵格式：{namespace}:{guildId}:{entityType}:{entityId}
 */
public interface CacheKeyGenerator {

  /** 鍵的命名空間前綴。 */
  String NAMESPACE = "cache";

  /**
   * 生成貨幣餘額的緩存鍵。
   *
   * @param guildId Discord 公會 ID
   * @param userId Discord 用戶 ID
   * @return 緩存鍵
   */
  String balanceKey(long guildId, long userId);

  /**
   * 生成遊戲代幣的緩存鍵。
   *
   * @param guildId Discord 公會 ID
   * @param userId Discord 用戶 ID
   * @return 緩存鍵
   */
  String gameTokenKey(long guildId, long userId);
}

File: packages/shared/src/infra/cache/cache-key-generator.ts
/**
 * Cache key generator interface providing consistent cache key format.
 * Key format: cache:{guildId}:{entityType}:{entityId}
 * Matches Java CacheKeyGenerator.
 *
 * NOTE: guildId and userId are typed as string rather than a numeric type because
 * discord.js uses snowflake strings for all IDs (e.g., "123456789012345678").
 * This is a JS/TS modeling choice — the underlying values are snowflake strings.
 */
export interface CacheKeyGenerator {
  readonly NAMESPACE: string;

  /** Generates a cache key for balance lookups. */
  balanceKey(guildId: string, userId: string): string;

  /** Generates a cache key for game token lookups. */
  gameTokenKey(guildId: string, userId: string): string;
}

/**
 * Default cache key generator implementation.
 * Key format:
 *   - balance:  cache:balance:{guildId}:{userId}
 *   - gametoken: cache:gametoken:{guildId}:{userId}
 */
export class DefaultCacheKeyGenerator implements CacheKeyGenerator {
  readonly NAMESPACE = 'cache';
  private static readonly ENTITY_BALANCE = 'balance';
  private static readonly ENTITY_GAME_TOKEN = 'gametoken';

  balanceKey(guildId: string, userId: string): string {
    return `${this.NAMESPACE}:${DefaultCacheKeyGenerator.ENTITY_BALANCE}:${guildId}:${userId}`;
  }

  gameTokenKey(guildId: string, userId: string): string {
    return `${this.NAMESPACE}:${DefaultCacheKeyGenerator.ENTITY_GAME_TOKEN}:${guildId}:${userId}`;
  }
}
